Develop

DenisSangi opened the pull request:
Sprint_2

DSangi committed:
add jacoco report

Commit:
Sprint_2 Flow_6

pomegranate282922 commented on src/test/java/com/example/LionHasManeExceptionTest.java:
```diff
@@ -0,0 +1,20 @@
+
+        Exception expectedException = new Exception("Используйте допустимые значения пола животного - самец или самка");
+
+        try {
```
Твой тест будет зелёным, если исключение НЕ выкинется, а это баг. Используй методы, доступные в junit4 для тестирования исключений. Например, вот https://jsparrow.github.io/rules/replace-j-unit-expected-exception.html#code-changes

pomegranate282922 commented on src/test/java/com/example/LionTest.java:
```diff
@@ -0,0 +1,34 @@
+    @Test
+    public void lionGetKittensTest() throws Exception {
+        //Arrange
+        Lion lion = spy(new Lion( "Самка"));
```
В чем смысл использования spy в этом случае? Почему не просто экземпляр lion?

pomegranate282922 commented on src/main/java/com/example/Lion.java:
```diff
@@ -1,32 +1,22 @@
-import java.util.List;
-
-public class Lion {
+public class Lion extends Feline {
```
Сделай, пожалуйста, изоляцию lion от feline через конструктор. То, что ты предлагаешь - это не изоляция, а изменение иерархии классов. 
и соответственно переделай тесты на Lion.

pomegranate282922 commented on src/test/java/com/example/CatGetFoodTest.java:
```diff
@@ -0,0 +1,31 @@
+
+
+public class CatGetFoodTest {
+
```
А в чем смысл использования именно spy? Чем недостаточен mock в данном случае?

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals(actual, expected);
+    }
+
+    @Test
```
Это странный тест, правда? Если у тебя есть тест, проверяющий, что этот метод с этим параметром вернёт ровно 2, то довольно логично предположить, что он не вернёт 7, не проверяя это)

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals(expected, actual);
+    }
+
+    @Test
```
Здесь то же самое. Нам не нужны тесты ради тестов. Их должно быть минимальное(!) количество из необходимых.

pomegranate282922 commented on src/test/java/com/example/FelineTest.java:
```diff
@@ -0,0 +1,80 @@
+        Feline feline = new Feline();
+        int expected = 1;
+        int actual = feline.getKittens();
+        System.out.println("actual = " + actual);
```
В чистовом варианте лучше убирай эти sout. Они только затрудняют чтение кода.

pomegranate282922 reviewed (commented):
Денис, давай разбираться.
1. Переделай изоляцию! Причём ты вроде бы уже сделал ее почти как надо в прошлых разах, а теперь пошёл не в ту сторону.
2. Везде, где можно - используй моки, а не spy. Spy - иногда бывает нужен, да. Но это не то же самое, что и мок, и если тебе нужно просто замокировать объект, не надо лепить spy. Прочитай про их разницу в Мокито.
3. Твои «негативные» тесты - не выдерживают критики) негативный сценарий - это не в смысле, что мы вызываем метод и проверяем, что результат не равен одному, второму, третьему значению ( при этом ты тут же уже проверил равенство ожидаемому). Это совсем про другое. Например, про то, что случится, если в параметры попадёт что-то неправильное.

Commit:
add jacoco report

Commit:
Second Try

О господи, Динара вот ты душишь. Я думал возьму академ и ревьюер поментяется. Твои разрабы наверное вешаются )))

pomegranate282922 commented on src/test/java/com/example/LionHasManeTest.java:
```diff
@@ -0,0 +1,41 @@
+    @Parameterized.Parameters
+    public static Object[] getLionData() {
+        return new Object[][]{
+                {feline,"Самец", true},
```
А зачем feline в параметризации - он же везде одинаковый!

pomegranate282922 commented on src/test/java/com/example/LionTest.java:
```diff
@@ -0,0 +1,44 @@
+
+        Lion lion = new Lion(feline, "Самец");
+        Mockito.when(feline.getFood("Хищник")).thenReturn(List.of("Животные", "Птицы", "Рыба"));
+        List<String> expectedListOfFood = List.of("Животные", "Птицы", "Рыба");
```
Дважды один и тот же список создаёшь.

pomegranate282922 reviewed (approved):
Денис, честно говоря, не очень приятно прочитать твой комментарий. Если мои замечания тебе видятся «душением» - аргументируй, пожалуйста. Я открыта к обсуждению, на некоторые вещи действительно можно по-разному смотреть и реализовывать.
Критичных замечаний у меня к работе нет. Зачёт.
Если есть такое желание у тебя - могу твои работы в следующих спринтах не брать, напиши в личку если так.
